File: number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        row, col = len(grid), len(grid[0])

        def dfs(x: int, y: int, route: List = []) -> List:
            if grid[x][y] == "0":
                return route

            route.append([x, y])
            grid[x][y] = "0"

            dx, dy = [1, 0, -1, 0], [0, -1, 0, 1]
            for n in range(4):
                nx, ny = x + dx[n], y + dy[n]
                if 0 <= nx < row and 0 <= ny < col and ([nx, ny]) not in route:
                    route = dfs(nx, ny, route)
            return route
        answer = 0

        for i in range(row):
            for j in range(col):
                answer += len(dfs(i, j, [])) > 0
        return answer

File: number-of-islands/number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int result = 0;
        Queue<int[]> queue = new LinkedList<>();
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == '1') {
                    result++;
                    queue.add(new int[]{i, j});
                    while (!queue.isEmpty()) {
                        int[] poll = queue.poll();
                        int x = poll[0], y = poll[1];
                        if (x >= 0 && x < grid.length && y >= 0 && y < grid[0].length && grid[x][y] == '1') {
                            grid[x][y] = '0';
                            queue.add(new int[]{x+1, y});
                            queue.add(new int[]{x, y+1});
                            queue.add(new int[]{x-1, y});
                            queue.add(new int[]{x, y-1});
                        }
                    }
                }
            }
        }
        return result;
    }
}